spec/fixtures/files/dummy_repo/repodata/repomd.xml

```diff
@@ -2,35 +2,35 @@
 <repomd xmlns="http://linux.duke.edu/metadata/repo" xmlns:rpm="http://linux.duke.edu/metadata/rpm">
  <revision>1500454273</revision>
 <data type="filelists">
-  <checksum type="sha256">837fb50abc9680b1e11e050901a56721855a5e854e85e46ceaad2c6816297e69</checksum>
+  <checksum type="sha256">67640a6a80cecbe77fe9bd3bc9381e017fd03433a6582a631cee6fc557cea354</checksum>
   <open-checksum type="sha256">a65752c29ed9e2921c1654327cae06cf58d0085ea76756ab240f2e7e42499654</open-checksum>
   <location href="repodata/837fb50abc9680b1e11e050901a56721855a5e854e85e46ceaad2c6816297e69-filelists.xml.gz"/>
   <timestamp>1500454273</timestamp>
-  <size>402</size>
+  <size>581</size>
   <open-size>929</open-size>
 </data>
 <data type="deltainfo">
-  <checksum type="sha256">a546b430098b8a3fb7d65493a9ce608fafcb32f451d0ce8bf85410191f347cc3</checksum>
+  <checksum type="sha256">8c9d9fbbe1b87009a145986a78ec1180dedc2455407f715d7a65174470d49793</checksum>
   <open-checksum type="sha256">e71433e2373d61a2e19f335c7f634b9bcfb1d458bc912626dc9ddc5f3e124981</open-checksum>
   <location href="repodata/a546b430098b8a3fb7d65493a9ce608fafcb32f451d0ce8bf85410191f347cc3-deltainfo.xml.gz"/>
```

Comment:
Hmm, seems like you've changed the `checksum` and `size`, but not the other fields: `location` still points to the file with the old name -- which shouldn't exist anymore, `open-checksum` and `open-size` probably should have changed as well.

So if this fixture still works -- I'm not sure why.

Reply:
Hi @ikapelyukhin, the fact that I updated metadata and checksums manually worried me from the very beginning. I just touched the files because there were no test scenarios for when the option `mirror_src == true` and when I removed the check for the flag on the parser classes, no tests broke (more details in d69b9c7), I needed some way to test assure it would behavior as expected.

I have little to no experience in building packages; I once wrote a RPM spec, a few years ago. I am trying to follow the [official OBS beginner's guide](https://openbuildservice.org/help/manuals/obs-beginners-guide/), but I am not acquainted with all the concepts, so I am struggling a bit to make a dummy repo myself.

Also, I have searched for a few other solutions; the command `createrepo` closely matches what I tried to do, but it does not generate the `deltainfo.xml` file.

Do you know a simpler way to build the metadata files or is it strictly necessary theses files matches all the requirements from a real repository?

Reply:
We can help with creating the repo. IIRC, I've built the RPMs with OBS and then created the repo with `createrepo`.

With regards to deltainfo:
```
createrepo -h | grep delta
  --deltas              create delta rpms and metadata
                        paths to look for older pkgs to delta against
```

There has to be more than one version of the package for deltas to be created, though.

Reply:
I tried running the `createrepo` command against the files in `spec/fixtures/files/dummy_repo`, after removing the `.drpm` files and the `repodata` directory. I changed directory to `dummy_repo` and executed `$ createrepo --deltas `.

The command generated all the `.xml.gz` files, but instead of a `deltainfo.xml.gz`, it generate a `prestodelta.xml.gz`, that was basically empty.

I took a look at the openSUSE documentation for [Standard RPM Metadata](https://en.opensuse.org/openSUSE:Standards_Rpm_Metadata#Current_Extensions) to see how to generate `deltainfo.xml` files. It seems it recomends an extension to the `createrepo` command called [`enhancerepo`](https://en.opensuse.org/openSUSE:Enhancerepo). I tried to install it from source and a [Ruby gem](https://rubygems.org/gems/enhancerepo) with the samename. However, I could not manage it to work correctly: the command tries to open a library file on `/usr/lib/rpm`, but on both openSUSE Leap 15.1 and Fedora 29 the path is for a directory.

Reply:
`prestodelta.xml.gz` -- that's Fedora flavor, the tags in the file are slightly different, too. Not sure why DRPMs aren't created, though. I don't have any recollection of how I created those DRPMs :thinking: 

In the meanwhile, I've created [home:ikapelyukhin:dummy_repo](https://build.opensuse.org/project/show/home:ikapelyukhin:dummy_repo). I'm not done fiddling with it yet, though, it takes a while for the repos to be published after the RPMs are built.

Reply:
Hi @ikapelyukhin, thank you very much for the time and support! Is it possible to also generate `.src.rpm` packages within OBS? The whole point of what I did (editing manually the fixtures files) was to have some way to test the condition `arch == `src` && mirror_src == true` when mirroring packages.

Currently, the project does not have fixture files for `.src.rpm` packages, so anything that

Reply:
@lcaparroz yes, you can see the source package was [built here](https://download.opensuse.org/repositories/home:/ikapelyukhin:/dummy_repo/openSUSE_Leap_15.1/src/).

It seems like publishing of build artifacts is rather slow on public OBS instance, I'm fiddling with project settings on our internal OBS instance to build a repo that has both source and deltarpm packages.

Reply:
@ikapelyukhin I managed to recreate the delta RPM packages with the `makedeltarpm` package (available in Fedora through the `deltarpm` package). It seems the command generate exactly the same file:

```sh
$ cd spec/fixtures/files/dummy_repo
$ makedeltarpm apples-0.1-0.x86_64.rpm apples-0.2-0.x86_64.rpm apples-0.2-0.x86_64.drpm

$ diff apples-0.1-0.x86_64.drpm oranges-0.1-0.x86_64.drpm; echo $?
Binary files apples-0.1-0.x86_64.drpm and oranges-0.1-0.x86_64.drpm differ
1

$ diff apples-0.1-0.x86_64.drpm apples-0.2-0.x86_64.drpm; echo $?
0
```

I did not find out a way to generate the `deltainfo.xml` file, though. Would OBS generate this automatically once there are two versions of a package? I read the [project configuration](https://build.opensuse.org/projects/home:ikapelyukhin:dummy_repo/prjconf) from your repository and it has this line `Repotype: rpm-md:deltainfo`, so I thought maybe OBS has the capability.

Reply:
@lcaparroz OBS definitely has the capacity, there's openSUSE/SLES drpms come from. :slightly_smiling_face: 
I've added a new version for `apples` package, hopefully drpm will get published in the project repo soon.

Reply:
Nice! About the VCR cassettes fixtures in the `rmt` project, for emulating requests to a repository, the tests run requests against `localhost`. I used `python3 -m http.server 80` to serve the files, but I don't know if there is a better way.

For what I understood, changing only the package/XML files makes no effects on the tests, since the VCR fixtures record the file contents as the request response body. I think it would be nice to have some documentation for how to record again the VCR cassettes and for generating package/XML files, since people new to the project may not know how to do this.

I can create the documentation for that, splitting it in two parts:

1. Generating RPM packages and XML metadata files;
1. Serving these files in `localhost` for recording the VCR cassettes.

What do you think about it? I could create a standalone markdown file and reference it in the README "Development Setup" section.

Reply:
Hi @ikapelyukhin, I replaced the files (RPM packages/XML metadata) I edited manually with the ones generated in your OBS repository, on commit 27ba069.

Just a few doubts:

* Does it make sense to have "src" delta RPMs packages (`*.src.drpm`)? I noticed that OBS just generated delta RPMs for the `x86_64` packages but not for `src` ones, and that the previous deltainfo XML parser (`RMT::Rpm::DeltainfoXmlParser`) had the rule for checking if a delta RPM package has `arch == 'src'` when `mirror_src == false`.
* The original fixture files included two versions for the same package (namely `0.0.1` and `0.0.2`), but after you committed a new version of the `apples` package to OBS, it is just serving the newer version and a delta RPM. Should we have both old and new versions for the fixtures?

Reply:
> About the VCR cassettes in the `rmt` project, for emulating requests to a repository, the tests run requests against `localhost`. I used `python3 -m http.server 80` to serve the files, but I don't know if there is a better way.

Any web server will do. I believe I've had a local nginx installation when I've recorded those, but `http.server` will do just as well.

> For what I understood, changing only the package/XML files makes no effects on the tests, since the VCR fixtures record the file contents as the request's response body. I think it would be nice to have some documentation for how to record again the VCR cassettes and for generating package/XML files, since people new to the project may not know how to do this.

The way the tests are currently done now -- yes, this is correct. Although it is possible to mirror a repository with a `file://`scheme URL pointed to the fixtures directory.

> I can create the documentation for that, splitting it in two parts:
> 
>     1. Generating RPM packages and XML metadata files;
> 
>     2. Serving these files in `localhost` for recording the VCR cassettes.
> 
> 
> What do you think about it? I could create a standalone markdown file and reference it in the README "Development Setup" section.

Sure, that sounds useful.

Reply:
> Hi @ikapelyukhin, I replaced the files (RPM packages/XML metadata) I edited manually with the ones generated in your OBS repository, on commit [27ba069](https://github.com/SUSE/rmt/commit/27ba0690485d81a8ca1480cb3c0cffb857e5fea2).
> 
> Just a few doubts:
> 
>     * Does it make sense to have "src" delta RPMs packages (`*.src.drpm`)? I noticed that OBS just generated delta RPMs for the `x86_64` packages but not for `src` ones, and that the previous deltainfo XML parser (`RMT::Rpm::DeltainfoXmlParser`) had the rule for checking if a delta RPM package has `arch == 'src'` when `mirror_src == false`.

In practice there are no source drpm packages.

> 
>     * The original fixture files included two versions for the same package (namely `0.0.1` and `0.0.2`), but after you committed a new version of the `apples` package to OBS, it is just serving the newer version and a delta RPM.

Yes, this is due to OBS project settings -- by default it keeps only the latest build artifact. I'm asking around to see if it's possible to set the project up to keep more that one version.

> Should we have both old and new versions for the fixtures?

The easiest solution to this perhaps would be to keep both repos and VCR cassettes -- the new one that you've recorded and the old one.

Reply:
@ikapelyukhin, I reverted the changes to the original fixtures and created another 'dummy repo' with the new files, recording a new VCR cassette (commit bd4fd5e).

Reply:
@ikapelyukhin, about the packages fixtures, do you think we have any other considerations?

The main point of this conversation was I fiddled with the metadata "by hand". Resuming, we followed the path below:

1. You generated the `.rpm`, `.drpm` and `.src` packages I needed for the `mirror src` test scenario;
1. I reverted the changes I did in the existing metadata, packages and VCR cassete;
1. Then I created a new VCR cassete using the newly generated packages/metadata.

Now, the original fixtures are not touched anymore and there's a test case covering the mentioned scenario.

Do you think having the documentation on how to generate these packages here to be strictly necessary at this point or could we postpone it until we have proper and richer information on this?

Reply:
@lcaparroz no, the documentation is not strictly necessary. `repomd.xml` in the new fixture now has:

```
  <tags>
    <repo>obsrepository://build.opensuse.org/home:ikapelyukhin:dummy_repo/openSUSE_Leap_15.1</repo>
    <repo>obsbuildid:1580990394</repo>
  </tags>
```

That should give a pretty good hint how this repo came to life. :slightly_smiling_face: